### DISCUSSION TITLE
Reusing parameters in a query with Enum values

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#7811] [Usage Questions] [answered]
Hi, I have a case where I have a large number of params in a fairly large query I am trying to run.

In most of the cases, my parameters are the same.
For example: 
```sql
SELECT *
FROM books
WHERE (genre = 'Fantasy' AND status = 'In progress')
  OR (genre = 'SciFi' AND status = 'In progress')
  OR (genre = 'Fiction' AND status = 'Completed')
```

So, I have an enum of values for the status - and I just have to repeat it multiple times based on the selections provided by users.

NOTE:
In the above example - I could change this to: `status = 'In progress' AND genre IN ('Fantasy', 'SciFi')` - but this is a lot more complex to do in my real queries.

If there some way for me to let sqlalchemy know that a variable I am giving in my `.filter()` is:

**Option 1**: Is a constant value which will be repeated 10 other times in the query, and so it can use the same named parameter for that ?

**Option 2**: A safe constant enum I am providing and hence does not require escaping (This may be a bit dicey security wise, so I would prefer option 1)

**Comment by CaselIT** [comment, marked as answer]
Hi,

You can use bindparam then set the values in the execute function.
See https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=bind#sqlalchemy.sql.expression.bindparam

**Reply by AbdealiLoKo** [reply]
Thanks for the pointer @CaselIT 

This seems like I will have to go a bit lower though.
Is there any option to do it with the ORM too ?

I currently do:
```python
session.query(Book).filter(
    or_(Book.genre == 'Fantasy', Book.status == 'In Progress'),
    or_(Book.genre == 'SciFi', Book.status == 'In Progress'),
    and_(Book.genre == 'Fiction', Book.status == 'Completed',
).all()
```

Would there be some nice way of doing:
```python
class Statuses(Enum):
    IN_PROG = bindvalue('In Progress').asparam('status_is_in_progress')
    COMPLETED = bindvalue('In Progress').asparam('status_is_complete')

session.query(Book).filter(
    or_(Book.genre == 'Fantasy', Book.status == Statuses.IN_PROG ),
    or_(Book.genre == 'SciFi', Book.status == Statuses.IN_PROG ),
    and_(Book.genre == 'Fiction', Book.status == Statuses.COMPLETED ),
).all()
```
This way, I can have unique names defined once - and reuse them through out my application
Instead of chasing after each scenario where I have such queries ?

**Reply by CaselIT** [reply]
You can do that with `bindparam`, just use the second argument `value`.

This should work
```py
class Statuses(Enum):
    IN_PROG = bindparam('status_is_in_progress', value='In Progress')
    COMPLETED = bindvalue('status_is_complete', value='Completed')
```

**Reply by AbdealiLoKo** [reply]
Thanks a lot @CaselIT for the super accurate and quick responses.
It's amazing that this has already been thought through and implemented in sqlalchemy :) 

Trying this out in my project today and will report back in case of further questions!
